internal/pkg/services/load-balancer/utils/utils.go

```diff
@@ -130,3 +133,100 @@ func GetTargetName(ctx context.Context, apiClient LoadBalancerClient, projectId,
+			usedCredentialsRefs = append(usedCredentialsRefs, *loadBalancer.Options.Observability.Metrics.CredentialsRef)
+		}
+	}
+
+	usedCredentialsMap := make(map[string]loadbalancer.CredentialsResponse)
+	for _, credential := range allCredentials {
+		if credential.CredentialsRef == nil {
+			continue
+		}
+		ref := *credential.CredentialsRef
+		if slices.Contains(usedCredentialsRefs, ref) {
+			usedCredentialsMap[ref] = credential
+		}
+	}
+
+	for _, credential := range usedCredentialsMap {
+		usedCredentialsSlice = append(usedCredentialsSlice, credential)
+	}
+
+	// sort credentials by reference to make output deterministic
+	sort.Slice(usedCredentialsSlice, func(i, j int) bool {
+		return *usedCredentialsSlice[i].CredentialsRef < *usedCredentialsSlice[j].CredentialsRef
+	})
+
+	return usedCredentialsSlice, nil
+}
+
+// GetUnusedObsCredentials returns a list of credentials that are not used by any load balancer for observability metrics or logs.
+// It compares the list of all credentials with the list of used credentials and returns a list of credentials that are not used.
+func GetUnusedObsCredentials(usedCredentials, allCredentials []loadbalancer.CredentialsResponse) []loadbalancer.CredentialsResponse {
+	var unusedCredentials []loadbalancer.CredentialsResponse
+	usedCredentialsRefs := make(map[string]bool)
+	for _, credential := range usedCredentials {
+		if credential.CredentialsRef != nil {
+			usedCredentialsRefs[*credential.CredentialsRef] = true
+		}
+	}
+
+	for _, credential := range allCredentials {
+		if credential.CredentialsRef == nil {
+			continue
+		}
+		if !usedCredentialsRefs[*credential.CredentialsRef] {
+			unusedCredentials = append(unusedCredentials, credential)
+		}
+	}
+
+	return unusedCredentials
+}
+
+// FilterCredentials filters a list of credentials based on the used and unused flags.
+// If used is true, it returns only the credentials that are used by load balancers for observability metrics or logs.
+// If unused is true, it returns only the credentials that are not used by any load balancer for observability metrics or logs.
+// If both used and unused are true, it returns an error.
+// If both used and unused are false, it returns the original list of credentials.
+func FilterCredentials(ctx context.Context, client LoadBalancerClient, credentials []loadbalancer.CredentialsResponse, projectId string, used, unused bool) ([]loadbalancer.CredentialsResponse, error) {
+	if !used && !unused {
+		return credentials, nil
+	}
+	if used && unused {
+		return nil, fmt.Errorf("used and unused flags are mutually exclusive")
+	}
+	filteredCredentials, err := GetUsedObsCredentials(ctx, client, credentials, projectId)
+	if err != nil {
+		return nil, fmt.Errorf("get used observability credentials: %w", err)
+	}
+
+	if unused {
+		filteredCredentials = GetUnusedObsCredentials(filteredCredentials, credentials)
+	}
+
+	return filteredCredentials, nil
+}
```

Comment:
Not sure if it's clearer, what do you think?

```suggestion
func FilterCredentials(ctx context.Context, client LoadBalancerClient, allCredentials []loadbalancer.CredentialsResponse, projectId string, used, unused bool) ([]loadbalancer.CredentialsResponse, error) {
	if !used && !unused {
		return allCredentials, nil
	}
	if used && unused {
		return nil, fmt.Errorf("used and unused flags are mutually exclusive")
	}
	usedCredentials, err := GetUsedObsCredentials(ctx, client, allCredentials, projectId)
	if err != nil {
		return nil, fmt.Errorf("get used observability credentials: %w", err)
	}

    if unused {
		return GetUnusedObsCredentials(filteredCredentials, allCredentials), nil
    }
    return usedCredentials, nil
}
```

Reply:
I think the formatting is a bit off in my suggestion

Reply:
I agree, it's better